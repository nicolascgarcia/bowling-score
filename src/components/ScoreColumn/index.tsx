import { useMemo } from 'react';
import { ScoreType } from '../../contexts/GameContext/gameContext.types';
import Container from '../Container';

type ScoreColumnProps = {
	currentRound: number;
	scoreIndex: number;
	currentPlayer: number;
	userIndex: number;
	scoreType: ScoreType;
	firstShot: number;
	secondShot: number;
	thirdShot: number | undefined;
	score: number;
};

export default function ScoreColumn({
	currentRound,
	scoreIndex,
	currentPlayer,
	userIndex,
	scoreType,
	firstShot,
	secondShot,
	thirdShot,
	score,
}: ScoreColumnProps) {
	const handleSecondShot = useMemo(() => {
		if (currentRound !== 9) {
			if (scoreType === ScoreType.NORMAL) {
				return secondShot;
			}
			if (scoreType === ScoreType.SPAIR) {
				return '/';
			}
			return 'X';
		}
		return secondShot;
	}, [currentRound, scoreType, secondShot]);

	return (
		<Container
			flex={1}
			display='flex'
			flexWrap='wrap'
			borderRightColor='gray'
			borderRightWidth='2px'
			borderRightStyle='solid'
			bg={
				currentRound === scoreIndex && currentPlayer === userIndex
					? '#8257e5'
					: undefined
			}
		>
			<Container
				width={1}
				display='flex'
				flexDirection='column'
				justifyContent='center'
				alignItems='center'
				borderBottomColor='gray'
				borderBottomWidth='2px'
				borderBottomStyle='solid'
			>
				<h1
					style={{
						textAlign: 'center',
						fontWeight: 'bold',
						color: '#fff',
					}}
				>
					{scoreIndex + 1}
				</h1>
				{scoreType !== ScoreType.NORMAL && (
					<Container
						bg={scoreType === ScoreType.STRIKE ? 'blue' : 'orange'}
						width={1}
						display='flex'
						justifyContent='center'
						py={1}
					>
						<span
							style={{
								color: '#fff',
								textTransform: 'uppercase',
								fontWeight: 'bold',
							}}
						>
							{scoreType === ScoreType.STRIKE ? 'strike' : 'spair'}
						</span>
					</Container>
				)}
			</Container>
			<Container flex={1} display='flex' flexDirection='column'>
				<Container display='flex' flexWrap='nowrap'>
					<Container width={1 / 2}>
						<h2
							style={{ textAlign: 'center', color: '#fff' }}
							aria-label={`firstshot-user${userIndex}-score${scoreIndex}`}
						>
							{firstShot}
						</h2>
					</Container>
					<Container
						width={1 / 2}
						display='flex'
						justifyContent='space-around'
						borderBottomColor='gray'
						borderBottomWidth='2px'
						borderBottomStyle='solid'
						borderLeftColor='gray'
						borderLeftWidth='2px'
						borderLeftStyle='solid'
						borderBottomLeftRadius='10px'
					>
						<h2
							aria-label={`secondshot-user${userIndex}-score${scoreIndex}`}
							style={{ textAlign: 'center', color: '#fff' }}
						>
							{handleSecondShot}
						</h2>
						{(thirdShot || thirdShot === 0) && (
							<h2 style={{ textAlign: 'center', color: '#fff' }}>
								{thirdShot}
							</h2>
						)}
					</Container>
				</Container>
				<Container
					width={1}
					flex={1}
					display='flex'
					flexDirection='column'
					justifyContent='center'
					alignItems='center'
				>
					<h2
						aria-label={`score-user${userIndex}-score${scoreIndex}`}
						style={{ color: '#fff' }}
					>
						{score}
					</h2>
				</Container>
			</Container>
		</Container>
	);
}
