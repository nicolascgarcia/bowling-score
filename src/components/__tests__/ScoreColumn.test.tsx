/* eslint-disable import/no-extraneous-dependencies */
import renderer from 'react-test-renderer';
import {
	Score,
	ScoreType,
	Player,
} from '../../contexts/GameContext/gameContext.types';
import ScoreColumn from '../ScoreColumn';

describe('<ScoreColumn/>', () => {
	it('should render ScoreColumn', () => {
		const scores: Score[] = new Array(9).fill(null).map(() => ({
			firstShot: 0,
			secondShot: 0,
			score: 0,
			shots: 0,
			type: ScoreType.NORMAL,
		}));

		scores.push({
			firstShot: 0,
			secondShot: 0,
			thirdShot: 0,
			score: 0,
			shots: 0,
			type: ScoreType.NORMAL,
		});

		const player: Player = {
			name: 'Nicolas',
			photo: '',
			scores,
			currentRound: 0,
			ended: false,
			hdcp: 0,
		};

		const component = renderer.create(
			<ScoreColumn
				currentPlayer={0}
				userIndex={0}
				currentRound={player.currentRound}
				firstShot={player.scores[0].firstShot}
				score={player.scores[0].score}
				scoreIndex={0}
				scoreType={player.scores[0].type}
				secondShot={player.scores[0].secondShot}
				thirdShot={player.scores[0].thirdShot}
			/>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
