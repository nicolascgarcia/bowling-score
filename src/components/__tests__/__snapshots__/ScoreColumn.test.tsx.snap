// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScoreColumn/> should render ScoreColumn 1`] = `
<div
  className="sc-bczRLJ jkldZp"
  display="flex"
>
  <div
    className="sc-bczRLJ jAFwIZ"
    display="flex"
    width={1}
  >
    <h1
      style={
        Object {
          "color": "#fff",
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      1
    </h1>
  </div>
  <div
    className="sc-bczRLJ gfvHoF"
    display="flex"
  >
    <div
      className="sc-bczRLJ heaOkQ"
      display="flex"
    >
      <div
        className="sc-bczRLJ eqsrV"
        width={0.5}
      >
        <h2
          aria-label="firstshot-user0-score0"
          style={
            Object {
              "color": "#fff",
              "textAlign": "center",
            }
          }
        >
          0
        </h2>
      </div>
      <div
        className="sc-bczRLJ fIFoXR"
        display="flex"
        width={0.5}
      >
        <h2
          aria-label="secondshot-user0-score0"
          style={
            Object {
              "color": "#fff",
              "textAlign": "center",
            }
          }
        >
          0
        </h2>
      </div>
    </div>
    <div
      className="sc-bczRLJ dbmcqy"
      display="flex"
      width={1}
    >
      <h2
        aria-label="score-user0-score0"
        style={
          Object {
            "color": "#fff",
          }
        }
      >
        0
      </h2>
    </div>
  </div>
</div>
`;
